### PR TITLE
Implement statfs for dragonfly, freebsd and openbsd

### PR DESCRIPTION
**Pull request by alesharik**

### CODE REVIEW

**Comment by semarie** [comment]
for OpenBSD, could you put the `c_uchar` definition in `src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs` ? the three definition you added are the same, so it could be factorized.

File: src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs [hunk 1]
```diff
@@ -25,8 +25,219 @@ s! {
...
+        pub f_owner: ::uid_t,
+        pub f_ctime: ::uint64_t,
+        pub f_fstypename: [::c_char; 16],
+        pub f_mntonname: [::c_char; 90],
+        pub f_mntfromname: [::c_char; 90],
+        pub f_mntfromspec: [::c_char; 90],
+        pub mount_info: mount_info,
+    }
+
+    pub union mount_info {
```
Comment (semarie) [inline src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs]: I have the following error will trying to run testsuite on openbsd:
```
error: no rules expected the token `union`
  --> src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs:54:9
   |
54 |     pub union mount_info {
   |         ^^^^^
```

The block is in `s!` marco block, and `union` seems unsupported by the macro

Comment (alesharik) [reply src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs]: I fixed unions. Can you rerun the testsuite?

**Comment by semarie** [comment]
- `c_uchar` is already globally defined : https://github.com/rust-lang/libc/blob/master/src/lib.rs#L131
- lot of errors related to union (I think your added support of union is only partial)

```
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13863:17: error: must use 'union' tag
 to refer to type 'mount_info'
cargo:warning=                mount_info* __test_field_type_statfs_mount_info(struct statfs* b) {
cargo:warning=                ^
cargo:warning=                union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13868:67: error: use of undeclared id
entifier 'mount_info'
cargo:warning=            uint64_t __test_size_mount_info(void) { return sizeof(mount_info); }
cargo:warning=                                                                  ^
...
cargo:warning=fatal error: too many errors emitted, stopping now [-ferror-limit=]
cargo:warning=20 errors generated.
exit code: 1
```

**Comment by alexcrichton** [comment]
Thanks! Looks like @semarie still has some errors though?

**Comment by alesharik** [comment]
Yes. And I'm building Rust for OpenBSD to fix them

**Comment by semarie** [comment]
@alesharik just a side note: you will have some unrelated errors regarding this PR (depending the openbsd version you use): `KERN_MAXID` on current, `SIGSTKSZ` (changed again recently) and `struct arphdr` alignment. I hope to be able to fix them soon, but you could take care of them too.

**Comment by semarie** [comment]
I still have the same kind of errors.

```
--- stdout
-----------------------------------------
cargo:rerun-if-changed=../src/lib.rs
cargo:rerun-if-changed=../src/macros.rs
cargo:rerun-if-changed=../src/dox.rs
cargo:rerun-if-changed=../src/unix/mod.rs
cargo:rerun-if-changed=../src/unix/bsd/mod.rs
cargo:rerun-if-changed=../src/unix/bsd/netbsdlike/mod.rs
cargo:rerun-if-changed=../src/unix/bsd/netbsdlike/openbsdlike/mod.rs
cargo:rerun-if-changed=../src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs
cargo:rerun-if-changed=../src/unix/bsd/netbsdlike/openbsdlike/openbsd/x86_64.rs
OPT_LEVEL = Some("0")
HOST = Some("x86_64-unknown-openbsd")
HOST = Some("x86_64-unknown-openbsd")
CC_x86_64-unknown-openbsd = None
CC_x86_64_unknown_openbsd = None
HOST_CC = None
CC = None
HOST = Some("x86_64-unknown-openbsd")
HOST = Some("x86_64-unknown-openbsd")
CFLAGS_x86_64-unknown-openbsd = None
CFLAGS_x86_64_unknown_openbsd = None
HOST_CFLAGS = None
CFLAGS = None
DEBUG = Some("true")
running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-m64" "-Wall" "-Wextra" "-Wall" "-Wextra" "-Werror" "-Wno-unused-parameter" "-Wno-type-limits" "-Wno-deprecated-declarations" "-Wno-deprecated-declarations" "-o" "/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.o" "-c" "/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c"
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13961:17: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                mount_info* __test_field_type_statfs_mount_info(struct statfs* b) {
cargo:warning=                ^
cargo:warning=                union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13966:67: error: use of undeclared identifier 'mount_info'
cargo:warning=            uint64_t __test_size_mount_info(void) { return sizeof(mount_info); }
cargo:warning=                                                                  ^
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13967:73: error: use of undeclared identifier 'mount_info'
cargo:warning=            uint64_t __test_align_mount_info(void) { return __alignof__(mount_info); }
cargo:warning=                                                                        ^
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13971:37: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    return offsetof(mount_info, ufs_args);
cargo:warning=                                    ^
cargo:warning=                                    union
cargo:warning=/usr/include/stddef.h:69:51: note: expanded from macro 'offsetof'
cargo:warning=#define offsetof(type, member)  __builtin_offsetof(type, member)
cargo:warning=                                                   ^
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13974:21: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    mount_info* foo = NULL;
cargo:warning=                    ^
cargo:warning=                    union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13979:72: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                struct ufs_args* __test_field_type_mount_info_ufs_args(mount_info* b) {
cargo:warning=                                                                       ^
cargo:warning=                                                                       union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13985:37: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    return offsetof(mount_info, mfs_args);
cargo:warning=                                    ^
cargo:warning=                                    union
cargo:warning=/usr/include/stddef.h:69:51: note: expanded from macro 'offsetof'
cargo:warning=#define offsetof(type, member)  __builtin_offsetof(type, member)
cargo:warning=                                                   ^
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13988:21: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    mount_info* foo = NULL;
cargo:warning=                    ^
cargo:warning=                    union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13993:72: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                struct mfs_args* __test_field_type_mount_info_mfs_args(mount_info* b) {
cargo:warning=                                                                       ^
cargo:warning=                                                                       union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:13999:37: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    return offsetof(mount_info, nfs_args);
cargo:warning=                                    ^
cargo:warning=                                    union
cargo:warning=/usr/include/stddef.h:69:51: note: expanded from macro 'offsetof'
cargo:warning=#define offsetof(type, member)  __builtin_offsetof(type, member)
cargo:warning=                                                   ^
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:14002:21: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    mount_info* foo = NULL;
cargo:warning=                    ^
cargo:warning=                    union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:14007:72: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                struct nfs_args* __test_field_type_mount_info_nfs_args(mount_info* b) {
cargo:warning=                                                                       ^
cargo:warning=                                                                       union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:14013:37: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    return offsetof(mount_info, iso_args);
cargo:warning=                                    ^
cargo:warning=                                    union
cargo:warning=/usr/include/stddef.h:69:51: note: expanded from macro 'offsetof'
cargo:warning=#define offsetof(type, member)  __builtin_offsetof(type, member)
cargo:warning=                                                   ^
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:14016:21: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    mount_info* foo = NULL;
cargo:warning=                    ^
cargo:warning=                    union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:14021:72: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                struct iso_args* __test_field_type_mount_info_iso_args(mount_info* b) {
cargo:warning=                                                                       ^
cargo:warning=                                                                       union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:14027:37: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    return offsetof(mount_info, msdosfs_args);
cargo:warning=                                    ^
cargo:warning=                                    union
cargo:warning=/usr/include/stddef.h:69:51: note: expanded from macro 'offsetof'
cargo:warning=#define offsetof(type, member)  __builtin_offsetof(type, member)
cargo:warning=                                                   ^
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:14030:21: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    mount_info* foo = NULL;
cargo:warning=                    ^
cargo:warning=                    union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:14035:80: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                struct msdosfs_args* __test_field_type_mount_info_msdosfs_args(mount_info* b) {
cargo:warning=                                                                               ^
cargo:warning=                                                                               union
cargo:warning=/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c:14041:37: error: must use 'union' tag to refer to type 'mount_info'
cargo:warning=                    return offsetof(mount_info, ntfs_args);
cargo:warning=                                    ^
cargo:warning=                                    union
cargo:warning=/usr/include/stddef.h:69:51: note: expanded from macro 'offsetof'
cargo:warning=#define offsetof(type, member)  __builtin_offsetof(type, member)
cargo:warning=                                                   ^
cargo:warning=fatal error: too many errors emitted, stopping now [-ferror-limit=]
cargo:warning=20 errors generated.
exit code: 1

--- stderr
thread 'main' panicked at '

Internal error occurred: Command "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-m64" "-Wall" "-Wextra" "-Wall" "-Wextra" "-Werror" "-Wno-unused-parameter" "-Wno-type-limits" "-Wno-deprecated-declarations" "-Wno-deprecated-declarations" "-o" "/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.o" "-c" "/data/semarie/repos/rust/build-libc/./build/debug/build/libc-test-12b6aa225dc34292/out/main.c" with args "cc" did not execute successfully (status code exit code: 1).

', /data/semarie/repos/rust/build-libc/./crates/registry/src/github.com-1ecc6299db9ec823/cc-1.0.17/src/lib.rs:2180:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Does `ctest` crate (which is used to generate the C code that will be used in the `libc` testsuite) have `union` support at all ?

**Comment by alesharik** [comment]
Nope, it doesn't. I'm working on it

**Comment by afdw** [comment]
Here is the PR that adds missing support for unions in the `ctest` crate: https://github.com/alexcrichton/ctest/pull/34

File: src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs [hunk 2]
```diff
@@ -25,8 +25,219 @@ s! {
...
+        pub cr_ngroups: ::c_short,
+        //https://github.com/openbsd/src/blob/master/sys/sys/syslimits.h#L44
+        pub cr_groups: [::gid_t; 16],
+    }
+
+    pub struct export_args {
+        pub ex_flags: ::c_int,
+        pub ex_root: ::uid_t,
+        pub ex_anon: xucred,
+        pub ex_addr: *mut sockaddr_in,
```
Comment (semarie) [inline src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs]: it is `struct sockaddr`

File: src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs [hunk 3]
```diff
@@ -25,8 +25,219 @@ s! {
...
+        pub ex_addrlen: ::c_int,
+        pub ex_mask: *mut sockaddr_in,
```
Comment (semarie) [inline src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs]: `struct sockaddr`

File: src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs [hunk 4]
```diff
@@ -25,8 +25,219 @@ s! {
...
+        pub ta_version: ::c_int,
+        pub ta_nodes_max: ::ino_t,
+        pub ta_size_max: ::off_t,
+        pub ta_root_uid: ::uid_t,
+        pub ta_root_gid: ::gid_t,
+        pub ta_root_mode: ::mode_t,
+    }
+
+    pub struct fusefs_args {
+        pub fspec: *mut ::c_char,
```
Comment (semarie) [inline src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs]: `pub name: *mut ::c_char;`

File: src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs [hunk 5]
```diff
@@ -25,8 +25,219 @@ s! {
...
+        pub mask: ::mode_t,
+        pub flags: ::c_int,
+    }
+
+    pub struct ntfs_args {
+        pub fspec: *mut ::c_char,
+        pub export_info: export_args,
+        pub uid: ::uid_t,
+        pub gid: ::gid_t,
+        pub mask: ::mode_t,
```
Comment (semarie) [inline src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs]: `pub mode: ::mode_t`

File: src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs [hunk 6]
```diff
@@ -25,8 +25,219 @@ s! {
...
+        pub ex_flags: ::c_int,
+        pub ex_root: ::uid_t,
+        pub ex_anon: xucred,
+        pub ex_addr: *mut sockaddr_in,
+        pub ex_addrlen: ::c_int,
+        pub ex_mask: *mut sockaddr_in,
+        pub ex_masklen: ::c_int,
+    }
+
+    pub struct in_addr {
```
Comment (semarie) [inline src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs]: `in_addr` is already defined

File: src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs [hunk 7]
```diff
@@ -25,8 +25,219 @@ s! {
...
+        pub s_addr: ::in_addr_t,
+    }
+    pub struct sockaddr_in {
```
Comment (semarie) [inline src/unix/bsd/netbsdlike/openbsdlike/openbsd/mod.rs]: `sockaddr_in` is already defined

**Comment by afdw** [comment]
@semarie yes, thanks. I have already fixed these problems here, in my fork: https://github.com/afdw/libc

**Comment by alexcrichton** [comment]
This is looking good to me! Are there any final bits to fix or is this good to go?

**Comment by afdw** [comment]
One problem that @semarie has mentioned is still not fixed yet:
```
cargo:warning=/home/user/test/libc/target/debug/build/libc-test-12b6aa225dc34292/out/main.c:14593:28: error: no member named 'mode' in 'msdosfs_args'
cargo:warning=                    return offsetof(struct msdosfs_args, mode);
cargo:warning=                           ^                             ~~~~
cargo:warning=/usr/include/stddef.h:69:32: note: expanded from macro 'offsetof'
cargo:warning=#define offsetof(type, member)  __builtin_offsetof(type, member)
cargo:warning=                                ^                        ~~~~~~
cargo:warning=/home/user/test/libc/target/debug/build/libc-test-12b6aa225dc34292/out/main.c:14597:40: error: no member named 'mode' in 'struct msdosfs_args'
cargo:warning=                    return sizeof(foo->mode);
cargo:warning=                                  ~~~  ^
cargo:warning=/home/user/test/libc/target/debug/build/libc-test-12b6aa225dc34292/out/main.c:14602:32: error: no member named 'mode' in 'struct msdosfs_args'
cargo:warning=                    return &b->mode;
cargo:warning=                            ~  ^
cargo:warning=/home/user/test/libc/target/debug/build/libc-test-12b6aa225dc34292/out/main.c:14681:28: error: no member named 'mask' in 'ntfs_args'
cargo:warning=                    return offsetof(struct ntfs_args, mask);
cargo:warning=                           ^                          ~~~~
cargo:warning=/usr/include/stddef.h:69:32: note: expanded from macro 'offsetof'
cargo:warning=#define offsetof(type, member)  __builtin_offsetof(type, member)
cargo:warning=                                ^                        ~~~~~~
cargo:warning=/home/user/test/libc/target/debug/build/libc-test-12b6aa225dc34292/out/main.c:14685:40: error: no member named 'mask' in 'struct ntfs_args'
cargo:warning=                    return sizeof(foo->mask);
cargo:warning=                                  ~~~  ^
cargo:warning=/home/user/test/libc/target/debug/build/libc-test-12b6aa225dc34292/out/main.c:14690:32: error: no member named 'mask' in 'struct ntfs_args'
cargo:warning=                    return &b->mask;
cargo:warning=                            ~  ^
```

**Comment by alesharik** [comment]
I think it is good to go

**Comment by afdw** [comment]
Yes, looks like it builds successfully and passes all tests.

**Comment by semarie** [comment]
it would be more simple to test if libc could be updated to use latest ctest crate... I need to rewrite the changes for each test

**Comment by semarie** [comment]
oh, it is just `Cargo.lock` that need to be updated

**Comment by semarie** [comment]
with an update of `Cargo.lock` to make ctest#bf780a0e be used,  it is ok for me: all tests passes on openbsd

**Pull request merged**